### PR TITLE
Update all calls to add/replace_grub_cmdline_settings

### PR DESCRIPTION
**Pull request by wjn740**
I modified all add/replace_grub_cmdline_settings function prototype.
It blocked some testings like 'security'. I update all 'calls' where that want to update
config in time.

@pevik @asmorodskyi @jouyingbin 

Signed-off-by: James Wang <jnwang@suse.com>

### CODE REVIEW

**Comment by pevik** [comment]
@wjn740 thanks, acked. But, can you, please, do one verify run? (it's obviously ok, but we've done some breaks in the past :)).
BTW: I was told by coolo in the past not to use `Signed-off-by: ` (os-autoinst projects are the only one where I don't use it).

File: tests/security/ima/evm_protection_hmacs.pm [hunk 1]
```diff
@@ -50,7 +50,7 @@ sub run {
     assert_script_run "setfattr -x security.ima $sample_app";
     validate_script_output "getfattr -m security.evm -d $sample_app", sub { m/^$/ };
 
-    replace_grub_cmdline_settings('evm=fix ima_appraise=fix', '', 1);
+    replace_grub_cmdline_settings('evm=fix ima_appraise=fix', '', update_grub => 1);
```
Comment (pevik) [inline tests/security/ima/evm_protection_hmacs.pm]: BTW I wonder if it work as 
`replace_grub_cmdline_settings('evm=fix ima_appraise=fix', update_grub => 1);`
i.e. having `$new` as `undef`. It'd probably introduce a warning, there would need to be added support for it in `change_grub_config()`. And explicitly set new as `''` is more readable.

**Comment by coolo** [comment]
> BTW: I was told by coolo in the past not to use `Signed-off-by: ` (os-autoinst projects are the only one where I don't use it).

Huh? Why would I say that?

**Comment by wjn740** [comment]
> @wjn740 thanks, acked. But, can you, please, do one verify run? (it's obviously ok, but we've done some breaks in the past :)).
I'm searching for their testcases to verifying.
> BTW: I was told by coolo in the past not to use `Signed-off-by: ` (os-autoinst projects are the only one where I don't use it).

OK.

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (wjn740) [reply tests/security/ima/evm_protection_hmacs.pm]: `
  1029 sub change_grub_config {
  1030     die((caller(0))[3] . ' expects from 3 to 5 arguments') unless (@_ >= 3 && @_ <= 5);
  1031     my ($old, $new, $search, $modifiers, $update_grub) = @_;
  1032     $modifiers   //= '';
  1033     $update_grub //= 0;
+ 1034     $new //= '';
  1035     $search = "/$search/" if defined $search;
  1036     
  1037     assert_script_run("sed -ie '${search}s/${old}/${new}/${modifiers}' " . GRUB_DEFAULT_FILE);
`

**Comment by pevik** [comment]
> > BTW: I was told by coolo in the past not to use `Signed-off-by: ` (os-autoinst projects are the only one where I don't use it).
> 
> Huh? Why would I say that?

You:
https://github.com/os-autoinst/openQA/pull/1182#issuecomment-273405543
Oliver:
https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/2250#issuecomment-268789913

Not worth of arguing (long time ago, maybe I didn't understand you at the time), I just stop adding `Signed-off-by:` then :)

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (pevik) [reply tests/security/ima/evm_protection_hmacs.pm]: Yes, that's exactly what I meant (+ mention it in perldoc and commit message).

**Comment by coolo** [comment]
ok, that wasnt' meant as a general rule though.

**Comment by pevik** [comment]
OK, thanks for explanation :)

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (wjn740) [reply tests/security/ima/evm_protection_hmacs.pm]: http://mitigations.qa2.suse.asia/tests/5372#step/sestatus/37

@jouyingbin Please take a look about the grub operation.

File: lib/bootloader_setup.pm [hunk 2]
```diff
@@ -1022,6 +1022,7 @@ sub grep_grub_cmdline_settings {
     change_grub_config($old, $new [, $search ] [, $modifiers ], [, $update_grub ]);
 
 Replace C<$old> with C<$new> in /etc/default/grub, using sed.
+C<$new> will be keep '' null string when it is undef.
```
Comment (pevik) [inline lib/bootloader_setup.pm]: I'd define it as: `C<$new>, default empty string`

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (pevik) [inline tests/security/ima/evm_protection_hmacs.pm]: I'd remove `'',` as it's now not needed :).

**Comment by lilyeyes** [comment]
Hi James, 
I have reviewed all "Security" related code changes,they looks good to me.
But, could you please offer the verify run otherwise I could not give you an "approval"?

BTW, it is close to sle12sp5 GMC hope this PR can be resolved asap.

**Comment by wjn740** [comment]
> Hi James,
> I have reviewed all "Security" related code changes,they looks good to me.
> But, could you please offer the verify run otherwise I could not give you an "approval"?
> 
> BTW, it is close to sle12sp5 GMC hope this PR can be resolved asap.

http://10.67.134.217/tests/5372#step/sestatus/37

**Comment by lilyeyes** [comment]
> > Hi James,
> > I have reviewed all "Security" related code changes,they looks good to me.
> > But, could you please offer the verify run otherwise I could not give you an "approval"?
> > BTW, it is close to sle12sp5 GMC hope this PR can be resolved asap.
> 
> http://10.67.134.217/tests/5372#step/sestatus/37
Thanks, the simple verify run is PASS.
Though I would like to see more verify runs since test case "SELinux" only verifies `set_*()` can work not include `replace_*()`.

 I would like to give you changes an "approval".

**Comment by asmorodskyi** [comment]
LGTM for me, just doubt about unanswered questions from @pevik . @wjn740 can you please take a look ?

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (wjn740) [reply tests/security/ima/evm_protection_hmacs.pm]: copy that.

**Comment by wjn740** [comment]
> LGTM for me, just doubt about unanswered questions from @pevik . @wjn740 can you please take a look ?

I missed it

File: lib/bootloader_setup.pm (hunk 2 above)
Comment (wjn740) [reply lib/bootloader_setup.pm]: done

**Comment by pevik** [comment]
@wjn740 `C<$new> will be keep '' null string when it is undef.` is a bit strange wording, I'd use something like `C<$new>, default empty string`.
Otherwise LGTM.

**Comment by wjn740** [comment]
> @wjn740 `C<$new> will be keep '' null string when it is undef.` is a bit strange wording, I'd use something like `C<$new>, default empty string`.
> Otherwise LGTM.

Sorry for my English. I have updated the code, now.

**Comment by wjn740** [comment]
There are more verifications:
http://10.67.134.217/tests/5387
http://10.67.134.217/tests/5388

**Comment by pevik** [comment]
@lilyeyes it haven't been merged yet, so nothing to be reverted yet :).

**Comment by wjn740** [comment]
> @lilyeyes it haven't been merged yet, so nothing to be reverted yet :).

She means revert 5362a6a52d4cfda3f6d9fedaaa2cf94845e6b281

commit 5362a6a52d4cfda3f6d9fedaaa2cf94845e6b281 (origin/jnwang/add_gurb_xen_support, jnwang/add_gurb_xen_support)
Author: James Wang <jnwang@suse.com>
Date:   Wed Oct 9 17:29:29 2019 +0800

    Fix "Add some API for Xen command-line" patch
    
    Fix compat_args function wasn't exported.

**Comment by lilyeyes** [comment]
> @lilyeyes it haven't been merged yet, so nothing to be reverted yet :).

Thanks for the replying I do not mean `#8727` I mean the PR before `#8727` (maybe it is `#8631` or the one before `#8631` )

**Comment by wjn740** [comment]
There is another verification:
http://10.67.134.217/tests/5385

**Comment by jouyingbin** [comment]
> There is another verification:
> http://10.67.134.217/tests/5385

I just check the verification run looks good to me,
The grub_mkconfig() can be executed correctly after appending update_grub => 1.
BTW, It will be great if you could collect all verification run in the top description.
Thanks for the updated.

**Comment by wjn740** [comment]
> > There is another verification:
> > http://10.67.134.217/tests/5385
> 
> I just check the verification run looks good to me,
> The grub_mkconfig() can be executed correctly after appending update_grub => 1.
> BTW, It will be great if you could collect all verification run in the top description.
> Thanks for the updated.

I don't know them before @lilyeyes told me. ;-)

File: tests/security/ima/evm_protection_hmacs.pm (hunk 1 above)
Comment (wjn740) [reply tests/security/ima/evm_protection_hmacs.pm]: @pevik Hi sir, why this ```''```, is not needed,
```
  1087 sub replace_grub_cmdline_settings {
  1088     my $old  = shift;
  1089     my $new  = shift;
  1090     my %args = testapi::compat_args(
  1091         {
  1092             old         => $old,
  1093             new         => $new,
  1094             update_grub => 0,
  1095             search      => get_cmdline_var(),
  1096         }, ['old', 'new', 'update_grub', 'search'], @_);
  1097     change_grub_config($old, $new, $args{search}, "g", $args{update_grub});
  1098 }

if remove it, line 1089, what will do ?
```

Comment (wjn740) [reply tests/security/ima/evm_protection_hmacs.pm]: sorry, Just a little bit confuse this in my brain, ...

**Pull request merged**